From aae383342ef4156f5b96d97b142c103c772de73a Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 4 May 2023 12:07:09 +0000
Subject: [PATCH 43/53] [Huawei] mm/sharepool: Reorganize sp_group_link_task()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Clear the spg_node operations and gather together the list update.
No new procedure introduced and just reorder the existing ones.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 103 +++++++++++++++++++++---------------------------
 1 file changed, 44 insertions(+), 59 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 1fa872e5d5a1..abc878c8ff50 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1192,15 +1192,15 @@ static struct sp_group *find_or_alloc_sp_group(int spg_id)
 }
 
 /* the caller must hold sp_global_sem */
-static struct sp_group_node *create_spg_node(struct mm_struct *mm,
+static struct sp_group_node *spg_node_alloc(struct mm_struct *mm,
 	unsigned long prot, struct sp_group *spg)
 {
 	struct sp_group_master *master = mm->sp_group_master;
 	struct sp_group_node *spg_node;
 
 	spg_node = kzalloc(sizeof(struct sp_group_node), GFP_KERNEL);
-	if (spg_node == NULL)
-		return ERR_PTR(-ENOMEM);
+	if (!spg_node)
+		return NULL;
 
 	INIT_LIST_HEAD(&spg_node->group_node);
 	INIT_LIST_HEAD(&spg_node->proc_node);
@@ -1209,26 +1209,9 @@ static struct sp_group_node *create_spg_node(struct mm_struct *mm,
 	spg_node->prot = prot;
 	meminfo_init(&spg_node->meminfo);
 
-	list_add_tail(&spg_node->group_node, &master->group_head);
-	master->group_num++;
-
 	return spg_node;
 }
 
-/* the caller must down_write(&spg->rw_lock) */
-static int insert_spg_node(struct sp_group *spg, struct sp_group_node *node)
-{
-	if (spg->proc_num + 1 == MAX_PROC_PER_GROUP) {
-		pr_err_ratelimited("add group: group reaches max process num\n");
-		return -ENOSPC;
-	}
-
-	spg->proc_num++;
-	list_add_tail(&node->proc_node, &spg->proc_head);
-
-	return 0;
-}
-
 /* the caller must down_write(&spg->rw_lock) */
 static void delete_spg_node(struct sp_group *spg, struct sp_group_node *node)
 {
@@ -1236,18 +1219,6 @@ static void delete_spg_node(struct sp_group *spg, struct sp_group_node *node)
 	spg->proc_num--;
 }
 
-/* the caller must hold sp_global_sem */
-static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
-	struct sp_group_node *spg_node)
-{
-	struct sp_group_master *master = mm->sp_group_master;
-
-	list_del(&spg_node->group_node);
-	master->group_num--;
-
-	kfree(spg_node);
-}
-
 /*
  * sp_group_link_task - Actually add a task into a group
  * @mm: specify the input task
@@ -1256,7 +1227,7 @@ static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
  *
  * The input @mm and @spg must have been initialized properly and could not
  * be freed during the sp_group_link_task().
- * the caller must hold sp_global_sem and down_write(&spg->rw_lock) in order.
+ * the caller must hold sp_global_sem.
  */
 static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
 			      unsigned long prot)
@@ -1265,31 +1236,49 @@ static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
 	struct sp_group_node *node;
 	struct sp_group_master *master = mm->sp_group_master;
 
-	if (is_process_in_group(spg, mm)) {
-		pr_err("task already in target group, id=%d\n", spg->id);
-		return -EEXIST;
-	}
-
 	if (master->group_num == MAX_GROUP_FOR_TASK) {
 		pr_err("task reaches max group num\n");
 		return -ENOSPC;
 	}
 
+	down_write(&spg->rw_lock);
+	if (is_process_in_group(spg, mm)) {
+		pr_err("task already in target group(%d)\n", spg->id);
+		ret = -EEXIST;
+		goto out_unlock;
+	}
+
+	node = spg_node_alloc(mm, prot, spg);
+	if (!node) {
+		ret = -ENOMEM;
+		goto out_unlock;
+	}
+
+	if (spg->proc_num + 1 == MAX_PROC_PER_GROUP) {
+		pr_err("add group: group(%d) reaches max process num\n", spg->id);
+		ret = -ENOSPC;
+		goto out_kfree;
+	}
+
 	ret = sp_group_setup_mapping(mm, spg);
 	if (ret)
-		return ret;
+		goto out_kfree;
 
-	node = create_spg_node(mm, prot, spg);
-	if (IS_ERR(node))
-		return PTR_ERR(node);
+	master->group_num++;
+	list_add_tail(&node->group_node, &master->group_head);
 
-	ret = insert_spg_node(spg, node);
-	if (unlikely(ret)) {
-		free_spg_node(mm, spg, node);
-		return ret;
-	}
+	spg->proc_num++;
+	list_add_tail(&node->proc_node, &spg->proc_head);
 
+	up_write(&spg->rw_lock);
 	return 0;
+
+out_kfree:
+	kfree(node);
+out_unlock:
+	up_write(&spg->rw_lock);
+
+	return ret;
 }
 
 /*
@@ -1418,17 +1407,19 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		goto out_put_mm;
 	}
 
-	down_write(&spg->rw_lock);
 	ret = sp_group_link_task(mm, spg, prot);
-	if (ret) {
-		up_write(&spg->rw_lock);
+	if (ret)
 		goto out_drop_group;
-	}
 
 	/*
 	 * create mappings of existing shared memory segments into this
 	 * new process' page table.
+	 *
+	 * Someone else may get the write lock before the following read lock, and
+	 * the spa in the spg may have changed. The newly created spa has been mapped
+	 * into the new task and we could get an errno of -EEXIST. Just ignore it.
 	 */
+	down_read(&spg->rw_lock);
 	for (p = rb_first(&spg->spa_root); p; p = rb_next(p)) {
 		spa = container_of(p, struct sp_area, spg_link);
 
@@ -1438,14 +1429,8 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 			ret = 0;
 		}
 	}
+	up_read(&spg->rw_lock);
 
-	up_write(&spg->rw_lock);
-
-	/*
-	 * to simplify design, we don't release the resource of
-	 * group_master and proc_stat, they will be freed when
-	 * process is exiting.
-	 */
 out_drop_group:
 	if (unlikely(ret)) {
 		up_write(&sp_global_sem);
@@ -1980,7 +1965,7 @@ static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 {
 	unsigned long addr = spa->va_start;
 	unsigned long size = spa_size(spa);
-	unsigned long flags = MAP_FIXED | MAP_SHARED | MAP_POPULATE |
+	unsigned long flags = MAP_FIXED_NOREPLACE | MAP_SHARED | MAP_POPULATE |
 			      MAP_SHARE_POOL;
 	unsigned long vm_flags = VM_NORESERVE | VM_SHARE_POOL | VM_DONTCOPY;
 	unsigned long pgoff = addr_offset(spa) >> PAGE_SHIFT;
-- 
2.17.1

